### ISSUE TITLE
Extend error reporting to cover SERVER_ERROR responses

### ISSUE THREAD

**Issue by DanTup** [Dart-Code/Dart-Code#213]
While trying out #211 I caused an error (the one from https://github.com/dart-lang/sdk/issues/27770) that came as a response to a request rather than as a notification:

```text
[17:48:54]: ==> {"id":"15","method":"analysis.updateContent","params":{"files":{"m:\\Coding\\TestStuff\\FootballGameDart-bad\\repro\\test\\team.dart":{"type":"change","edits":[{"offset":1812,"length":0,"replacement":"{}","id":""}]}}}}
[17:48:54]: <== {"error":{"code":"SERVER_ERROR","message":"The null object does not have a method '_subFromInteger'.\n\nNoSuchMethodError: method not found: '_subFromInteger'\nReceiver: null\nArguments: [2057]","stackTrace":"#0      Object._noSuchMethod (dart:core-patch/object_patch.dart:44)\n#1      Object.noSuchMethod (dart:core-patch/object_patch.dart:47)\n#2      int.- (dart:core-patch/integers.dart:17)\n#3      _UpdateElementOffsetsVisitor.visitElement (package:analyzer/src/task/incremental_element_builder.dart:754)\n#4      GeneralizingElementVisitor.visitVariableElement (package:analyzer/dart/element/visitor.dart:178)\n#5      GeneralizingElementVisitor.visitPropertyInducingElement (package:analyzer/dart/element/visitor.dart:168)\n#6      GeneralizingElementVisitor.visitFieldElement (package:analyzer/dart/element/visitor.dart:111)\n#7      FieldElementImpl.accept (package:analyzer/src/dart/element/element.dart:4241)\n#8      IncrementalCompilationUnitElementBuilder._replaceNode (package:analyzer/src/task/incremental_element_builder.dart:516)\n#9      IncrementalCompilationUnitElementBuilder._processClassMembers (package:analyzer/src/task/incremental_element_builder.dart:324)\n#10     IncrementalCompilationUnitElementBuilder._processUnitMembers (package:analyzer/src/task/incremental_element_builder.dart:463)\n#11     IncrementalCompilationUnitElementBuilder.build (package:analyzer/src/task/incremental_element_builder.dart:140)\n#12     AnalysisContextImpl._sourceChanged (package:analyzer/src/context/context.dart:1845)\n#13     AnalysisContextImpl.handleContentsChanged (package:analyzer/src/context/context.dart:1064)\n#14     AnalysisServer.updateContent.<anonymous closure>.<anonymous closure> (package:analysis_server/src/analysis_server.dart:1360)\n#15     List.forEach (dart:core-patch/growable_array.dart:258)\n#16     AnalysisServer.updateContent.<anonymous closure> (package:analysis_server/src/analysis_server.dart:1353)\n#17     _HashVMBase&MapMixin&&_LinkedHashMapMixin...
```

This is not caught by this error handling; we should see whether we can extend it.

Some errors are "expected" (eg. failed to reformat a file if it's invalid), so probably we only want to handle `SERVER_ERROR` here.

**Comment by devoncarew** [comment]
Ah, it's difficult to discriminate between responses that include  analysis server exceptions and just async analysis server exception notifications.

**Comment by DanTup** [comment]
>  it's difficult to discriminate between responses that include analysis server exceptions and just async analysis server exception notifications

Surely the opposite is true? It's because they're different that the new reporting only covered one? We already handle request responses that are errors (because we call `reject` instead of `fulfil` for the `.then` to work nicely), I think we can just inject there (if the request errored and the type is `SERVER_ERROR`).

I'll try it out when I get some time (likely Friday).

**Comment by devoncarew** [comment]
Ah, yeah, I didn't notice that some server responses have a field set for `SERVER_ERROR`; makes sense now. They're both a error response to the request and an indication that the server itself is at fault.

**Comment by DanTup** [comment]
@devoncarew I've implemented this (wasn't actually much effort). I added a new subscription to the `analyzer` for `SERVER_ERROR` error responses and in analyzer_status_reporter subscribed to it in the same way as your existing errors and just mapped `RequestError` onto a `ServerError`.

This seems to do what I wanted; other errors (eg. try to format when there's syntax errors) still fall through the original path, only `SERVER_ERROR` goes through this.

If you have a chance to review these changes, feedback is welcome.

FYI - I also added analytics so we get counts of server errors that could be useful (not tagged against this Issue).
